Sources/SourceKit/SourceKitServer.swift

```diff
@@ -338,54 +335,63 @@ extension SourceKitServer {
   // MARK: - Text synchronization
 
   func openDocument(_ note: Notification<DidOpenTextDocument>, workspace: Workspace) {
-    workspace.documentManager.open(note)
+    workspace.documentManager.open(note.params)
 
     let textDocument = note.params.textDocument
     workspace.buildSettings.registerForChangeNotifications(for: textDocument.url)
 
     if let service = languageService(for: textDocument.url, textDocument.language, in: workspace) {
-      service.send(note.params)
+      service.openDocument(note.params)
     }
   }
 
   func closeDocument(_ note: Notification<DidCloseTextDocument>, workspace: Workspace) {
-    workspace.documentManager.close(note)
+    workspace.documentManager.close(note.params)
 
     let url = note.params.textDocument.url
     workspace.buildSettings.unregisterForChangeNotifications(for: url)
 
     if let service = workspace.documentService[url] {
-      service.send(note.params)
+      service.closeDocument(note.params)
     }
   }
 
   func changeDocument(_ note: Notification<DidChangeTextDocument>, workspace: Workspace) {
-    workspace.documentManager.edit(note)
+    workspace.documentManager.edit(note.params)
 
     if let service = workspace.documentService[note.params.textDocument.url] {
-      service.send(note.params)
+      service.changeDocument(note.params)
     }
   }
 
   func willSaveDocument(_ note: Notification<WillSaveTextDocument>, workspace: Workspace) {
-
+    if let service = workspace.documentService[note.params.textDocument.url] {
+      service.willSaveDocument(note.params)
+    }
   }
 
   func didSaveDocument(_ note: Notification<DidSaveTextDocument>, workspace: Workspace) {
-
+    if let service = workspace.documentService[note.params.textDocument.url] {
+      service.didSaveDocument(note.params)
+    }
   }
 
   // MARK: - Language features
 
   func completion(_ req: Request<CompletionRequest>, workspace: Workspace) {
-    toolchainTextDocumentRequest(
-      req,
-      workspace: workspace,
-      fallback: CompletionList(isIncomplete: false, items: []))
+    guard let service = workspace.documentService[req.params.textDocument.url] else {
+      req.reply(CompletionList(isIncomplete: false, items: []))
+      return
+    }
+    service.completion(req)
   }
 
   func hover(_ req: Request<HoverRequest>, workspace: Workspace) {
-    toolchainTextDocumentRequest(req, workspace: workspace, fallback: nil)
+    guard let service = workspace.documentService[req.params.textDocument.url] else {
```

Comment:
While `toolchainTextDocumentRequest` no longer makes sense, maybe we should pull the lookup into a method `languageService<PositionRequest: TextDocumentRequest>(for request: PositionRequest)`, since I could see us changing this behaviour and wanting to keep all of them in sync.

Reply:
Done, but what do you think instead of having a `registerToolchainTextDocumentRequest(handler, fallbackValue)` to register these?

Reply:
Could you expand a bit on what you mean? I'm not sure what exactly you have in mind.

Reply:
Something like

```
  func registerToolchainTextDocumentRequest<PositionRequest: TextDocumentRequest>(
    _ fallback: PositionRequest.Response,
    _ requestHandler: @escaping (SourceKitServer) -> (Request<PositionRequest>, ToolchainLanguageServer) -> Void
  ) {
    _register { [unowned self] (req: Request<PositionRequest>) in
      guard let workspace = self.workspace else {
        return req.reply(.failure(.serverNotInitialized))
      }
      guard let languageService = workspace.documentService[request.params.textDocument.url] else {
        return req.reply(fallback)
      }

      requestHandler(self)(req, languageService )
    }
  }
```

Reply:
Implemented this